src/foam/dao/CSVSink.js

```diff
@@ -76,34 +106,99 @@ foam.CLASS({
             Null: function(value) {}
           }, function(value) {
             this.output_(value.toString());
-          })
+        }),
+      javaCode: `
+        if ( value instanceof String ) {
+          getSb().append("\\"");
+          getSb().append(((String)value).replace("\\"", "\\"\\""));
+          getSb().append("\\"");
+        } else if ( value instanceof Number ) {
+          getSb().append(value.toString());
+        } else if ( value instanceof Boolean ) {
+          getSb().append(value.toString());
+        } else if ( value instanceof Date ) {
+          getSb().append(value.toString());
+        } else if ( value instanceof FObject ) {
+          output_(value.toString());
+        } else if ( value instanceof List ) {
+          output_(value.toString());
+        } else {
+          getSb().append(value.toString());
+        }
+      `
     },
-
-    function newLine_() {
-      this.csv += '\n';
-      this.isNewLine = true;
+    {
+      name: 'newLine_',
+      code: function() {
+        this.csv += '\n';
+        this.isNewLine = true;
+      },
+      javaCode: `
+        getSb().append("\\n");
+        setIsNewLine(true);
+      `
     },
+    {
+      name: 'eof',
+      javaCode: 'setCsv(getSb().toString());'
+    },
+    {
+      name: 'put',
+      code: function(obj) {
+        if ( ! this.of ) this.of = obj.cls_;
+        var element = undefined;
+        if ( ! this.isHeadersOutput ) {
+          this.props.forEach((name) => {
```

Comment:
I would have done isFirstRow and put the header outputting into its own function.

`if ( this.isFirstRow ) this.outputHeaders`

Not a big deal